src/main/java/io/jenkins/plugins/extratoolinstallers/installers/VersionCheckingToolInstaller.java

```diff
@@ -0,0 +1,280 @@
+     */
+    @CheckForNull
+    public String getVersionMax() {
+        return versionMax;
+    }
+
+    /**
+     * Sets {@link #getVersionMax()}.
+     * 
+     * @param versionMax New value.
+     */
+    @DataBoundSetter
+    public void setVersionMax(String versionMax) {
+        this.versionMax = versionMax;
+    }
+
+    @Override
+    public FilePath performInstallation(@Nonnull ToolInstallation tool, @Nonnull Node node,
+            @CheckForNull TaskListener log) throws IOException, InterruptedException {
+        final FilePath executablePath = findExecutableOnNodeOrThrow(node, log);
+        final FilePath parent = executablePath.getParent();
+        if (parent == null) {
+            // This shouldn't happen, hence not localized.
+            throw new IllegalStateException(
+                    "Executable " + executablePath + " has no parent folder");
+        }
+        final String relPathOrNull = getRelativePath();
+        final FilePath resultToReturn;
+        if (relPathOrNull == null || relPathOrNull.equals(".")) {
+            resultToReturn = parent;
+        } else {
+            resultToReturn = parent.child(relPathOrNull);
+        }
+        final String[] vCmd = getVersionCmd();
+        final Pattern vPattern = getVersionPattern();
+        final String vMax = getVersionMax();
+        final String vMin = getVersionMin();
+        if (vCmd != null && vPattern != null && (vMin != null || vMax != null)) {
+            final ByteArrayOutputStream output = new ByteArrayOutputStream();
+            final Launcher launcher = node.createLauncher(log);
+            runCommandOnNode(launcher, resultToReturn, vCmd, output);
+            final String cmdOutput = output.toString(StandardCharsets.UTF_8.name());
+            final String parsedVersion = VersionChecker.parseVersionCmdOutputForVersion(vPattern, cmdOutput);
+            final int versionComparisonResult = VersionChecker.checkVersionIsInRange(vMin, vMax, parsedVersion);
+            if (versionComparisonResult != 0) {
+                throw new WrongVersionException(executablePath.getRemote(), resultToReturn.getRemote(), parsedVersion, vMin,
+                        vMax);
+            }
+        }
+        return resultToReturn;
+    }
+
+    @Nonnull
+    abstract FilePath findExecutableOnNodeOrThrow(@Nonnull Node node, @CheckForNull final TaskListener logOrNull) throws IOException, InterruptedException;
+
+    // package access for test purposes only
+    @Restricted(NoExternalUse.class)
+    void runCommandOnNode(final Launcher launcher, final FilePath pwd, final String[] cmd,
+            final OutputStream output) throws IOException, InterruptedException {
+        launcher.launch().cmds(cmd).stdout(output).pwd(pwd).join();
+    }
+
+    // package access for test purposes only
+    @Restricted(NoExternalUse.class)
+    @Nonnull
+    abstract FilePath.FileCallable<String> mkCallable(@CheckForNull final TaskListener logOrNull);
+}
```

Comment:
This class should contain a Descriptor subclass that contains the common code from the subclasses' Descriptor classes (so they don't have to repeat the code).

I don't think it'll need an `@Extension` annotation (that's only for the non-abstract subclasses), and it can/should be abstract.